### PR TITLE
Implement more robust DSL Spec2 Next

### LINKED ISSUE #24: Error when running `crystal spec`
When running `crystal spec` the following error output is given:

`Assertion failed: ((i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"), function init, file /tmp/llvm20151214-69554-hagc7s/llvm-3.6.2.src/lib/IR/Instructions.cpp, line 281.
fish: 'crystal run spec/spec2_spec.cr' terminated by signal SIGABRT (Abort)`

@waterlink  I'm currently trying to implement #23, can you help me with this?

### PR DESCRIPTION
**Pull request by waterlink**
fixes #24 

This PR is a full rewrite of DSL, that drives this library. No public API should have been changed.

@marceloboeira Could you try this branch out on one of your projects (with `branch: slight-rewrite` in `shard.yml` respective dependency).

I believe that current implementation should be more robust in terms of breakage by compiler update. This still abuses macros heavily, but much less than the previous version did.

### CODE REVIEW

File: spec/matchers_spec.cr [hunk 1]
```diff
@@ -494,12 +494,14 @@ Spec2.describe Spec2::Matchers do
     end
 
     context "when type does not match" do
-      expect {
-        expect(42).to be_a(String)
-      }.to raise_error(
-        Spec2::ExpectationNotMet,
-        "Expected 42 to be a String",
-      )
+      it "fails" do
```
Comment (waterlink) [inline spec/matchers_spec.cr:497]: Apparently, we had a bug in DSL here, where it actually permitted expectations out of scope of example.

File: spec/spec2_spec.cr [hunk 2]
```diff
@@ -1,521 +0,0 @@
-require "./spec_helper"
```
Comment (waterlink) [inline spec/spec2_spec.cr:1]: This file is fully re-written in a unit-test fashion and sits in `./unit/simple_dsl_spec.cr`

File: unit/simple_dsl_spec.cr [hunk 3]
```diff
@@ -0,0 +1,418 @@
+require "./spec_helper"
```
Comment (waterlink) [inline unit/simple_dsl_spec.cr:1]: This is tested by stdlib's `spec`. It makes it easier to understand what is happening, when DSL itself breaks.

**Comment by marceloboeira** [comment]
@waterlink Thanks!

I've tested the repository spec, as also in one of my projects. Everything seems to be working perfectly.

With this I'm able to take a good luck on implementing #23 and #25 :dancers:

**Comment by waterlink** [comment]
@marceloboeira Great. Then I am merging this.

**Comment by waterlink** [comment]
@marceloboeira OK, I have released version `0.4.2`, since it shouldn't have changed any behavior.

**Comment by marceloboeira** [comment]
@waterlink that seems correct :+1:

**Pull request merged**